transport/peer/peerlist/roundrobin.go

```diff
@@ -0,0 +1,181 @@
+package peerlist
+
+import (
+	"context"
+	"sync"
+
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/internal/errors"
+
+	"github.com/uber-go/atomic"
+)
+
+// NewRoundRobin creates a new round robin PeerList using
+func NewRoundRobin(peerIDs []transport.PeerIdentifier, agent transport.PeerAgent) (transport.PeerList, error) {
+	rr := &roundRobin{
+		peerToNode: make(map[string]*roundRobinNode, len(peerIDs)),
+		agent:      agent,
+		started:    atomic.NewBool(false),
+		nextNode:   nil,
+	}
+
+	err := rr.createPeers(peerIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	return rr, nil
+}
+
+type roundRobin struct {
+	sync.Mutex
+
+	peerToNode map[string]*roundRobinNode
+	agent      transport.PeerAgent
+	started    *atomic.Bool
+	nextNode   *roundRobinNode
```

Comment:
Instead of implementing your own, you can use `list.List`.

```
struct {
    index map[string]*list.Element
    peers *list.List
    next  *list.Element
}
```

`ChoosePeer` is simply,

    peer := pl.next.(transport.Peer)
    pl.next = pl.next.Next()
    return peer

`addPeer`:

    elem := pl.peers.PushBack(p)
    pl.index[p.Identifier()] = elem

`removePeer`:

    elem := pl.index[p.Identifier()]
    pl.peers.Remove(elem)
    delete(pl.index, p.Identifier())